include/picongpu/plugins/output/images/Visualisation.hpp

```diff
@@ -157,286 +159,654 @@ struct typicalFields < 5 >
+                DataSpace< simDim > const cellIdx = DataSpaceOperations< simDim >::template map< SuperCellSize >( linearIdx );
 
+                DataSpace< DIM2 > const localCell(
+                    cellIdx[ transpose.x() ],
+                    cellIdx[ transpose.y() ]
+                );
 
-        // counter is always DIM2
-        typedef DataBox < PitchedBox< float_X, DIM2 > > SharedMem;
-        sharedMemExtern( shBlock, float_X );
+                if( isImageThreadCtx[ idx ] )
+                {
+                    counter( localCell ) = float_X(0.0);
+                }
+            }
+        );
 
-        const DataSpace<simDim> blockSize(blockDim);
-        SharedMem counter(PitchedBox<float_X, DIM2 > ((float_X*) shBlock,
-                                                      DataSpace<DIM2 > (),
-                                                      blockSize[transpose.x()] * sizeof (float_X)));
+        // wait that shared memory  is set to zero
+        __syncthreads();
 
-        if (isImageThread)
-        {
-            counter(localCell) = float_X(0.0);
-        }
+        using FramePtr = typename T_ParBox::FramePtr;
+        FramePtr frame = pb.getFirstFrame( suplercellIdx );
 
+        // each virtual worker works on a particle in the frame
+        ForEachIdx< ParticleDomCfg > forEachParticle( workerIdx );
 
-        if (localId == 0)
+        while( frame.isValid( ) )
         {
-            frame = pb.getFirstFrame(block);
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    auto particle = frame[ linearIdx ] ;
+                    if( particle[ multiMask_ ] == 1)
+                    {
+                        int const linearCellIdx = particle[ localCellIdx_ ];
+                        // we only draw the first slice of cells in the super cell (z == 0)
+                        DataSpace< simDim > const particleCellOffset(
+                            DataSpaceOperations< simDim >::template map< SuperCellSize >( linearCellIdx )
+                        );
+                        bool const isParticleOnSlice = IsPartOfSlice< >{}(
+                            particleCellOffset + supercellCellOffset,
+                            sliceDim,
+                            localDomainOffset,
+                            slice
+                        );
+                        if( isParticleOnSlice )
+                        {
+                            DataSpace< DIM2 > const reducedCell(
+                                particleCellOffset[ transpose.x( ) ],
+                                particleCellOffset[ transpose.y( ) ]
+                            );
+                            atomicAdd(
+                                &( counter( reducedCell ) ),
+                                particle[ weighting_ ] / particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE,
```

Comment:
OPTIONAL:
wouldn't it be more efficient if the normalization `.../particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE` is done once after the summation via atomicAdd instead of per particle?
(Of course this is only true if there are less workers than particles per frame, which is however the standard case in 2D simulations).

Reply:
for example in line 642

Reply:
no wee need to normalize it here to avoid bad precision if the weighting is large

Reply:
Okay, do you mean we might run into the maximum (limit) of the float number? Because division does not change precision, and summation errors will occur independent of the rescaling.

Reply:
Yes we can pass the maximum for float and the addition will have a large error because of the huge difference of the long sum.